src/tools/aml-language-server/handlers.ts

```diff
@@ -0,0 +1,138 @@
+      capabilities: {
+        // Tell the client that the server works in FULL text document sync mode
+        textDocumentSync: TextDocumentSyncKind.Full,
+        hoverProvider: true,
+        signatureHelpProvider: {
+          triggerCharacters: ['(', ','],
+        },
+        definitionProvider: true,
+        typeDefinitionProvider: true,
+        referencesProvider: true,
+        documentSymbolProvider: true,
+        workspaceSymbolProvider: true,
+        xworkspaceReferencesProvider: true,
+        xdefinitionProvider: true,
+        xdependenciesProvider: true,
+        completionProvider: {
+          resolveProvider: true,
+          triggerCharacters: ['.'],
+        },
+        codeActionProvider: true,
+        renameProvider: true,
+        executeCommandProvider: {
+          commands: [],
+        },
+        xpackagesProvider: true,
+      }
+    };
+    return result;
+  },
+
+  textDocumentCompletion: async (
+      params, context: AmlServiceContext):
+      Promise<MessageTypes.CompletionItem> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Completing for : ${uri}...`);
+        return undefined;
+      },
+
+  textDocumentDidSave: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Handling save for: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      },
+
+  textDocumentDidChange: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Handling changes for: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      },
+
+  textDocumentDidOpen: async (params, context: AmlServiceContext):
+      Promise<NotificationMessage> => {
+        const uri = (params as {textDocument: {uri: string}}).textDocument.uri;
+        context.logger.info(`Opened: ${uri}...`);
+        return publishDiagnostics(uri, context);
+      }
+};
+
+async function publishDiagnostics(uri: string, context: AmlServiceContext):
+    Promise<NotificationMessage> {
+      const diagnosticParams: PublishDiagnosticsParams =
+          await gatherDiagnostics(uri, context);
+      return {
+        jsonrpc,
+        method: 'textDocument/publishDiagnostics',
+        params: diagnosticParams
+      };
+    }
+
+async function gatherDiagnostics(uri: string, context: AmlServiceContext):
+    Promise<PublishDiagnosticsParams> {
+      const path = uri2path(normalizeUri(uri));
+      // TODO(cypher1): Catch exception and list them for later.
+      const errors: ManifestError[] = [];
+      try {
+        const manifest = await Manifest.load(path, new LspLoader(context));
+        errors.splice(0, 0, ...Manifest.getErrors(manifest));
```

Comment:
errors.push(...Manifest.getErrors(manifest)); ?

Reply:
Thanks (I had wrongly assumed that would only work on a 2D array).